oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/client/JdbcRegisteredClientRepository.java

```diff
@@ -0,0 +1,257 @@
+		this.jdbcOperations = jdbcOperations;
+		this.registeredClientRowMapper = this::defaultRegisteredClientRowMapper;
+		this.registeredClientParameterMapper = this::defaultRegisteredClientParameterMapper;
+		this.listParser = this::defaultListParse;
+		this.listFormatter = this::defaultListFormat;
+	}
+
+	/**
+	 * Allows changing of {@link RegisteredClient} row mapper implementation
+	 *
+	 * @param registeredClientRowMapper mapper implementation
+	 */
+	public void setRegisteredClientRowMapper(RowMapper<RegisteredClient> registeredClientRowMapper) {
+		Assert.notNull(registeredClientRowMapper, "registeredClientRowMapper cannot be null");
+		this.registeredClientRowMapper = registeredClientRowMapper;
+	}
+
+	/**
+	 * Allows changing of SQL parameter mapper for {@link RegisteredClient}
+	 *
+	 * @param registeredClientParameterMapper mapper implementation
+	 */
+	public void setRegisteredClientParameterMapper(Function<RegisteredClient, List<SqlParameterValue>> registeredClientParameterMapper) {
+		Assert.notNull(registeredClientParameterMapper, "registeredClientParameterMapper cannot be null");
+		this.registeredClientParameterMapper = registeredClientParameterMapper;
+	}
+
+	/**
+	 * Allows changing format of list attributes of {@link RegisteredClient}
+	 *
+	 * @param listParser new list parser function
+	 */
+	public void setListParser(Function<String, Collection<String>> listParser) {
+		Assert.notNull(listParser, "listParser cannot be null");
+		this.listParser = listParser;
+	}
+
+	/**
+	 * Allows changing format of list attributes of {@link RegisteredClient}
+	 *
+	 * @param listFormatter new list formatter function
+	 */
+	public void setListFormatter(Function<Collection<String>, String> listFormatter) {
+		Assert.notNull(listFormatter, "listFormatter cannot be null");
+		this.listFormatter = listFormatter;
+	}
+
+	@Override
+	public void save(RegisteredClient registeredClient) {
+		Assert.notNull(registeredClient, "registeredClient cannot be null");
+		RegisteredClient foundClient = this.findBy("id = ? OR client_id = ? OR client_secret = ?",
+				registeredClient.getId(), registeredClient.getClientId(), registeredClient.getClientSecret());
+
+		if (null != foundClient) {
+			Assert.isTrue(!foundClient.getId().equals(registeredClient.getId()),
+					"Registered client must be unique. Found duplicate identifier: " + registeredClient.getId());
+			Assert.isTrue(!foundClient.getClientId().equals(registeredClient.getClientId()),
+					"Registered client must be unique. Found duplicate client identifier: " + registeredClient.getClientId());
+			Assert.isTrue(!foundClient.getClientSecret().equals(registeredClient.getClientSecret()),
+					"Registered client must be unique. Found duplicate client secret for identifier: " + registeredClient.getId());
+		}
+
+		List<SqlParameterValue> parameters = this.registeredClientParameterMapper.apply(registeredClient);
+		PreparedStatementSetter pss = new ArgumentPreparedStatementSetter(parameters.toArray());
```

Comment:
In conjunction with adding a `lob` value for `client_secret`, this would need to be switched to using a `LobHandler` and `LobCreatorArgumentPreparedStatementSetter`. See [JdbcOAuth2AuthorizedClientService](https://github.com/spring-projects/spring-security/blob/9d2db8983807cbe29b2efb5f6ee3c17e5b05364d/oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/JdbcOAuth2AuthorizedClientService.java#L191-L195) for an example.

Reply:
Reimplemented it as described above, yet I don't like it at all. It seems that just for having `client_secret` field of arbitrary length, considerable bunch of complexity was added. In addition local class LobCreatoArgumentPreparedStatement is now duplicated (plus another copy in spring-security proper), so it is probably a candidate to factor out as independent class. Yet it does not solve main problem: do we need all that LOB handling complexity just to skip declaring limit of `client_secret` field ? Change has been pushed, so please decide if I should simplify it or leave as it is.

Reply:
Feedback received, @rlewczuk, and thanks for being flexible on this one. I will log an issue to reduce code duplication for the setter implementation.